lib/tools/broccoli/node-modules-dest-copy.ts

```diff
@@ -0,0 +1,84 @@
+import shelljs = require("shelljs");
+import {wrapBroccoliPlugin} from './broccoli-plugin-wrapper-factory';
+import constants = require("./../../constants");
+
+/**
+ * Intercepts each directory as it is copied to the destination tempdir,
+ * and tees a copy to the given path outside the tmp dir.
+ */
+export class DestCopy implements IBroccoliPlugin {
+  constructor(private inputPath: string, private cachePath: string, private outputRoot: string, private projectDir: string) {}
+
+  public rebuild(treeDiff: IDiffResult): void {	  
+	let dependencies = this.getDependencies();	  
+	let devDependencies = this.getDevDependencies(this.projectDir);
+	
+	treeDiff.changedDirectories.forEach(changedDirectory => {
+		let changedDirectoryAbsolutePath = path.join(this.inputPath, constants.NODE_MODULES_FOLDER_NAME, changedDirectory);
+		let packageJsonFilePath = path.join(changedDirectoryAbsolutePath, "package.json");
+		let fileContent = require(packageJsonFilePath);
+	
+		if(!devDependencies[fileContent.name]) { // Don't flatten dev dependencies
+	
+			let currentDependency = {
+				name: fileContent.name,
+				version: fileContent.version,
+				directory: path.dirname(packageJsonFilePath)
+			};
+			
+			let addedDependency = dependencies[currentDependency.name];
+			if (addedDependency) {
+				if (semver.gt(currentDependency.version, addedDependency.version)) {
+					let currentDependencyMajorVersion = semver.major(currentDependency.version);
+					let addedDependencyMajorVersion = semver.major(addedDependency.version);
+	
+					let message = util.format("Old version %s, new version %s. Old location %s, new location %s.", addedDependency.version, currentDependency.version, addedDependency.directory, currentDependency.directory);
+					currentDependencyMajorVersion === addedDependencyMajorVersion ? console.log(message) : console.log(message);
+	
+					dependencies[currentDependency.name] = currentDependency;
+				}
+			} else {
+				dependencies[currentDependency.name] = currentDependency;
+			}
+		}
+	});
+	
+	_.each(dependencies, dependency => shelljs.cp("-R", dependency.directory, path.join(this.outputRoot, path.basename(dependency.directory))));
+	
+	// Cache input tree
+	let projectFilePath = path.join(this.projectDir, constants.PACKAGE_JSON_FILE_NAME);	
+	let projectFileContent = require(projectFilePath);
+	projectFileContent[constants.NATIVESCRIPT_KEY_NAME][constants.NODE_MODULE_CACHE_PATH_KEY_NAME] = this.inputPath;	
+	fs.writeFileSync(projectFilePath, JSON.stringify(projectFileContent, null, "\t"), { encoding: "utf8" });	
+  }
+  
+  private getDependencies(): IDictionary<any> {
+	  let result = Object.create(null);
+	  try {
+		  let dirs = fs.readdirSync(this.outputRoot);
+		  _.each(dirs, dir => {
+			  let fileContent = require(path.join(dir, constants.PACKAGE_JSON_FILE_NAME));
+			  result[fileContent.name] = fileContent;
+			 });
+	  } catch(err) {
+		  result = Object.create(null);
+	  }
+	  
+	  return result;
```

Comment:
As the function is written, it silently skips all dependencies in case any of them is corrupted. This means that code after that will randomly fail. Maybe it is better to let the exception cancel the current operation so that the user can have contextual information what's wrong and can fix it.